ui/src/app/views/breadcrumb/component.ts

```diff
@@ -0,0 +1,111 @@
+import {Component, OnDestroy, OnInit} from '@angular/core';
+import {ActivatedRoute, NavigationEnd, PRIMARY_OUTLET, Router} from '@angular/router';
+import {Subscription} from 'rxjs/Subscription';
+
+export interface Breadcrumb {
+  label: string;
+  url: string;
+}
+@Component({
+  selector: 'app-breadcrumb',
+  templateUrl: './component.html',
+  styleUrls: ['../../styles/buttons.css',
+    '../../styles/cards.css',
+    '../../styles/headers.css',
+    '../../styles/inputs.css',
+    './component.css']
+})
+export class BreadcrumbComponent implements OnInit, OnDestroy {
+  subscription: Subscription;
+  breadcrumbs: Breadcrumb[];
+  constructor(
+      private activatedRoute: ActivatedRoute,
+      private router: Router) {}
+
+  /**
+   * Generate a breadcrumb using the default label and url. Uses the route's
+   * paramMap to do any necessary variable replacement. For example, if we
+   * have a label value of ':wsid' as defined in a route's breadcrumb, we can
+   * do substitution with the 'wsid' value in the route's paramMap.
+   *
+   * @param {String} label
+   * @param {String} url
+   * @param {ActivatedRoute} route
+   * @returns {Breadcrumb}
+   */
+  private static makeBreadcrumb(
+      label: string,
+      url: string,
+      route: ActivatedRoute
+  ): Breadcrumb {
+    let newLabel = label;
+    // Perform variable substitution in label only if needed.
+    if (newLabel.indexOf(':') >= 0) {
+      try {
+        const paramMap = route.snapshot.paramMap;
+        for (const k of paramMap.keys) {
+          newLabel = newLabel.replace(':' + k, paramMap.get(k));
+        }
+      } catch (e) {
```

Comment:
What are you expecting to here? Missing key? Would probably prefer improved precondition checks rather than having a catch-all.